### ISSUE TITLE
TODO list for ImmutableGraph related features

### ISSUE THREAD

**Issue by zheng-da** [dmlc/dgl#369]
## 🚀 Feature

Current `ImmutableGraph` (merged in #342 ) still lacks some functionalities. This might be a problem when user switch from normal graph to it (some APIs will break). For example, ImmutableGraph needs to support FindEdges, EdgeSubgraphs, batching graphs, etc.

We use this issue to book-keep all the missing features.

**Comment by yzh119** [comment]
Could you elaborate more on EdgeSubgraphs?

**Comment by yzh119** [comment]
TODOs:
- FindEdges
1. `FindEdge(dgl_id_t eid)`
2. `FindEdges(IdArray eids)`

**Comment by yzh119** [comment]
@zheng-da I've implemented the two APIs: 
```cpp
std::pair<dgl_id_t, dgl_id_t> ImmutableGraph::FindEdge(dgl_id_t eid) const {
  auto out_csr = GetOutCSR();
  // The edge_ids must have been sorted.
  auto ptr = std::lower_bound(out_csr->edge_ids.begin(), out_csr->edge_ids.end(), eid);
  //auto ptr = std::find(out_csr->edge_ids.begin(), out_csr->edge_ids.end(), eid);
  CHECK(eid == *ptr);
  const int64_t idx = ptr - out_csr->edge_ids.begin();
  dgl_id_t col = out_csr->indices[idx];
  dgl_id_t row = std::lower_bound(out_csr->indptr.begin(), out_csr->indptr.end(), idx) - out_csr->indptr.begin();
  return std::pair<dgl_id_t, dgl_id_t>(row, col);
}

ImmutableGraph::EdgeArray ImmutableGraph::FindEdges(IdArray eids) const {
  dgl_id_t* eid_data = static_cast<dgl_id_t*>(eids->data);
  const auto len = eids->shape[0];
  std::vector<dgl_id_t> src, dst;
  for (size_t k = 0; k < len; k++) {
    auto edge = ImmutableGraph::FindEdge(eid_data[k]);
    src.push_back(edge.first);
    dst.push_back(edge.second);
  }

  IdArray rst_src = IdArray::Empty({len}, eids->dtype, eids->ctx);
  IdArray rst_dst = IdArray::Empty({len}, eids->dtype, eids->ctx);
  dgl_id_t* rst_src_data = static_cast<dgl_id_t*>(rst_src->data);
  dgl_id_t* rst_dst_data = static_cast<dgl_id_t*>(rst_dst->data);
  std::copy(src.begin(), src.end(), rst_src_data);
  std::copy(dst.begin(), dst.end(), rst_dst_data);
  return ImmutableGraph::EdgeArray{rst_src, rst_dst, eids};
}
```
For efficient implementation the `edge_ids` must be sorted, which might not be the general case, even though we may accomplish this by enforcing the rows and cols in coo-matrix for initialization to be sorted during the graph building phase(this is the current solution I used for beam search in Transformer).

I was wondering if we could first map the real `edge_ids` to a virtual `edge_ids` via a hash map, where the virtual `edge_ids` was guaranteed to be sorted. The additional cost here is to store a hash mapping in the CSR class.

**Comment by zheng-da** [comment]
If we store a hashtable, it's better to store an edge list. Unfortunately, we should add many data structures to immutable graph index because we expect it'll be used to store giant graphs. We hope it uses as little memory as possible to store the graph structure.

**Comment by yzh119** [comment]
Merged in #404
